### PR TITLE
t/ckeditor5/1514: Created a guide with solutions for common conversion problems reported by the users

### PR DESCRIPTION
**Pull request by oleq**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Docs: Created a guide with solutions for common conversion problems reported by the users. Closes ckeditor/ckeditor5#1514.

---

### Additional information

Requires https://github.com/ckeditor/ckeditor5/issues/1520 to get the link UI working in the examples upon clicking a link.

#### Some comments

* The number of examples is limited. OTOH they are very time–consuming to prepare.
* The guide does not cover lists and tables. A complex conversion is used there I didn't have time to research.
* I feel that even with an introduction, this guide is very hard to read and understand. It asks for a proper "comprehensive conversion guide" but it is a huuuuuuge topic.
    * If we write a comprehensive guide in the future, we should move the intro part from this guide there and merge it.

### CODE REVIEW

**Comment by Reinmar** [comment]
More ideas – there are some really cool examples in https://github.com/ckeditor/ckeditor5-core/blob/poc/catch-all-converters/tests/manual/converters.js and https://github.com/ckeditor/ckeditor5-core/blob/poc/catch-all-converters/tests/manual/converters.html. We need to add them to the list :)

Also... we need tests for the converters in these samples :((( But I have an idea how to do that ;>

**Comment by jodator** [comment]
I'll add those to the PR.

**Comment by oleq** [comment]
The same thing as d370243 may be necessary for other snippets too. They work without it but it could be an accident that during `writer.wrap()` the result is the old element that had the custom prop already defined.

**Comment by jodator** [comment]
@Reinmar & @oleq:  I've forgot to push the second example with fonts. I think that we can have that one only from the two that were created as they seems too similar to me. But I'm happy to add the section with converting font-names also if you see it helpful.

File: docs/_snippets/framework/extending-content-add-heading-class.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ClassicEditor
+	.create( document.querySelector( '#snippet-heading-class' ), {
+		cloudServices: CS_CONFIG,
+		extraPlugins: [ AddClassToAllHeading1 ],
+		toolbar: {
+			viewportTopOffset: window.getViewportTopOffsetConfig()
+		}
+	} )
+	.then( editor => {
+		window.editor = editor;
```
[the remark this answers is not in the document]
Comment (oleq) [reply docs/_snippets/framework/extending-content-add-heading-class.js]: Yeah. Probably `window.editors = { ... }`.

**Comment by Reinmar** [comment]
One more case to cover: divs with classes: 

```js
  editor.model.schema.register( 'div', {
      allowWhere: '$block',
      allowContentOf: '$block', // May also be `$root` if it's supposed to behave as a wrapper.
      allowAttributes: 'class'
  } );

  // Model-to-view convert for the div element (attrbiutes are converted separately).
  editor.conversion.elementToElement( {
      model: 'div',
      view: 'div'
  } );

  editor.conversion.attributeToAttribute( {
      model: { name: 'div', key: 'class' },
      view: 'class'
  } );
```

**Comment by Reinmar** [comment]
Also, we actually have to review many of those examples now because of link decorators. I wouldn't necessary remove them, but explain that this can also be achieved with that feature.

**Comment by jodator** [comment]
I've added:

- small fixes to the docs (mainly written by me)
- references to the link decorators

Examples cross-linked to the link decorators:

- Adding a CSS class to inline elements
- Adding an HTML attribute to certain inline elements
- Adding a CSS class to certain inline elements
- Loading content with a custom attribute

The code works AFAICS and the examples are still relevant but some of them are already implemented as link decoreators so we might think in the future to rewrite them to similar cases but withouth using `< href="...">`.

Still the whole guide is just needed to be published IMO :)

**Comment by Reinmar** [comment]
Great job, @jodator! Thanks!

I guess we still have some things that we could add to this guide. We've been doing some PoC's in ckeditor5-core, e.g. Could you also open a followup ticket with links to those and some other ideas that we posted here or that come to your head? You've been implementing many strange cases for our customers, so there must be something :D

**Comment by jodator** [comment]
@Reinmar most of the ideas from this PR and a ticket are implemented in the guide. Anyway, the follow up is here: https://github.com/ckeditor/ckeditor5-engine/issues/1761 as I couldn't find more ideas.

File: docs/_snippets/framework/extending-content-arbitrary-attribute-values.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/* globals ClassicEditor, console, window, document */
+
+import { CS_CONFIG } from '@ckeditor/ckeditor5-cloud-services/tests/_utils/cloud-services-config';
+import Font from '@ckeditor/ckeditor5-font/src/font';
```
Comment (Reinmar) [inline docs/_snippets/framework/extending-content-arbitrary-attribute-values.js]: Can't add a plugin to a build:

image

File: docs/_snippets/framework/extending-content-custom-figure-attributes.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/* globals ClassicEditor, console, window, document */
+
+import { CS_CONFIG } from '@ckeditor/ckeditor5-cloud-services/tests/_utils/cloud-services-config';
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
```
Comment (Reinmar) [inline docs/_snippets/framework/extending-content-custom-figure-attributes.js]: Same here – you can't add plugins like this.

Comment (jodator) [reply docs/_snippets/framework/extending-content-custom-figure-attributes.js]: Fixed.

**Comment by Reinmar** [comment]
6th attempt at reviewing this :D

**Comment by Reinmar** [comment]
OK, I admit I didn't check the code samples themselves but rather focused on gluing together some intro and making this guide easier to follow by splitting it. Let's just polish it with time.

Anyway, thanks for all the content guys. Lots of valuable information will finally be available to the people.

**Pull request merged**